### PR TITLE
Donut iobinding

### LINKED ISSUE #1196: Add IOBinding for ORTModelForVision2Seq
### Feature request

Add IOBinding support for ORTModelForVision2Seq

### Motivation

Priority model for inference

### PR DESCRIPTION
**Pull request by IlyasMoutawwakil**
# What does this PR do?

Fixes #1196 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by IlyasMoutawwakil** [comment]
I will check if this change interferes with TrOCR

**Comment by IlyasMoutawwakil** [comment]
@mht-sharma it seems TrOCR doesn't support `use_cache` and `use_io_binding`. What was the reason for that ?

**Comment by mht-sharma** [comment]
> @mht-sharma it seems TrOCR doesn't support `use_cache` and `use_io_binding`. What was the reason for that ?

Not sure about the IOBinding, but the TRocr does not output the `past-key-values` when setting `use_cache= True ` maybe something odd in the modelling code. That's, why `use_cache=True` throws error.

**Comment by IlyasMoutawwakil** [comment]
Okay, since they're coupled I will investigate that as part of this pr.
Also tell me what you think of the method I added and if it's what you had in mind.

**Comment by fxmarty** [comment]
@IlyasMoutawwakil I think donut-mbart is untested with `ORTModelForVision2Seq` (only vit-gpt2 is), it could be useful to add it in the tests!

**Comment by IlyasMoutawwakil** [comment]
@fxmarty thanks I will add them

**Comment by mht-sharma** [comment]
@IlyasMoutawwakil could you check if the same shape can work for all encoders (because this may not be the case). Maybe trying "nlpconnect/vit-gpt2-image-captioning", "naver-clova-ix/donut-base-finetuned-docvqa" and trocr.

**Comment by IlyasMoutawwakil** [comment]
"nlpconnect/vit-gpt2-image-captioning" seems to take too long (doesn't stop) during merging. unmerged model fails with io binding in the decoder, @mht-sharma do you have any idea why ?
`Error in execution: Non-zero status code returned while running Transpose node. Name:'/decoder/transformer/h.1/attn/Transpose_1' Status Message: /onnxruntime_src/onnxruntime/core/framework/execution_frame.cc:171 onnxruntime::common::Status onnxruntime::IExecutionFrame::GetOrCreateNodeOutputMLValue(int, int, const onnxruntime::TensorShape*, OrtValue*&, const onnxruntime::Node&) shape && tensor.Shape() == *shape was false. OrtValue shape verification failed. Current shape:{1,12,197,64} Requested shape:{1,12,1,64}`

**Comment by mht-sharma** [comment]
> "nlpconnect/vit-gpt2-image-captioning" seems to take too long (doesn't stop) during merging. unmerged model fails with io binding in the decoder, @mht-sharma do you have any idea why ?
> `Error in execution: Non-zero status code returned while running Transpose node. Name:'/decoder/transformer/h.1/attn/Transpose_1' Status Message: /onnxruntime_src/onnxruntime/core/framework/execution_frame.cc:171 onnxruntime::common::Status onnxruntime::IExecutionFrame::GetOrCreateNodeOutputMLValue(int, int, const onnxruntime::TensorShape*, OrtValue*&, const onnxruntime::Node&) shape && tensor.Shape() == *shape was false. OrtValue shape verification failed. Current shape:{1,12,197,64} Requested shape:{1,12,1,64}`

From error the issue seems to be more of mismatched IOBinding shapes. would check on my end also

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 1]
```diff
@@ -383,6 +388,33 @@ def forward(
...
+        elif self.normalized_config.config.model_type == "donut-swin":
+            # for donut-swin models
+            encoder_sequence_length = (
+                self.normalized_config.config.image_size[0]
+                * self.normalized_config.config.image_size[1]
+                // self.normalized_config.config.hidden_size
+            )
+        else:
+            raise ValueError(
+                f"Unsupported encoder model type {self.normalized_config.config.model_type} for VisionEncoderDecoder."
```
Comment (JingyaHuang) [inline optimum/onnxruntime/modeling_seq2seq.py:406]: Can you also mention in the error message what types are supported?

**Review by JingyaHuang — APPROVED** [review]
LGTM, thanks a lot for working on this @IlyasMoutawwakil !

File: tests/onnxruntime/utils_onnxruntime_tests.py [hunk 2]
```diff
@@ -99,6 +99,7 @@
     "xlm_roberta": "hf-internal-testing/tiny-xlm-roberta",
     "vision-encoder-decoder": "hf-internal-testing/tiny-random-VisionEncoderDecoderModel-vit-gpt2",
     "trocr": "microsoft/trocr-small-handwritten",
+    "donut": "fxmarty/tiny-doc-qa-vision-encoder-decoder",
```
Comment (fxmarty) [inline tests/onnxruntime/utils_onnxruntime_tests.py:102]: Can you move it up so that the dict is ordered?

File: tests/onnxruntime/test_modeling.py [hunk 3]
```diff
@@ -3920,8 +3920,7 @@ def test_compare_generation_to_io_binding(
 
 
 class ORTModelForVision2SeqIntegrationTest(ORTModelTestMixin):
```
Comment (fxmarty) [inline tests/onnxruntime/test_modeling.py:3922]: There seem to be no test for logits match with/without IO Binding in this class, can you add it?

File: optimum/onnxruntime/modeling_seq2seq.py (hunk 1 above)
Comment (fxmarty) [inline optimum/onnxruntime/modeling_seq2seq.py:406]: This is an other case where we may benefit from subclassing, I'll try to do a PR rather soon than later.

Comment (IlyasMoutawwakil) [reply optimum/onnxruntime/modeling_seq2seq.py]: done!

File: tests/onnxruntime/test_modeling.py (hunk 3 above)
Comment (IlyasMoutawwakil) [reply tests/onnxruntime/test_modeling.py:3922]: done! but I had to add a model type `donut` to tasks manager, even though it's not exactly a model type. is that okay in this case ?

Comment (IlyasMoutawwakil) [reply tests/onnxruntime/test_modeling.py:3922]: I guess tests are passing now (on linux). I had some problems with `ORT_CUDA_UNAVAILABLE`. I can perfectly run models with `CUDAExeutionProvider` but the pipelines provider check fails because of `ORT_CUDA_UNAVAILABLE`.

Comment (fxmarty) [reply tests/onnxruntime/test_modeling.py:3922]: Are you sure the model is indeed dispatched to CUDA EP? You can check with

```python
session_options = onnxruntime.SessionOptions()
session_options.log_severity_level = 0

ort_model = ORTModel.from_pretrained(
    ...,
    session_options=session_options
)
```

**Comment by fxmarty** [comment]
@IlyasMoutawwakil What is missing for this PR to be merged?

**Comment by IlyasMoutawwakil** [comment]
@fxmarty nothing, I believe it's ready to merge

**Comment by IlyasMoutawwakil** [comment]
@fxmarty except windows, everything else is passing. there's also a quota issue.

**Review by fxmarty — APPROVED** [review]
LGTM thanks a lot!
